src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/Admin/EditTypePart.cshtml

```diff
@@ -1,15 +1,33 @@
 @model EditTypePartViewModel
+@using System
 @using OrchardCore.ContentManagement.Metadata.Settings
 @using OrchardCore.DisplayManagement;
 
 @inject OrchardCore.DisplayManagement.Theming.IThemeManager ThemeManager
 @inject OrchardCore.DisplayManagement.Descriptors.IShapeTableManager ShapeTableManager
+@inject IEnumerable<IShapeBindingResolver> ShapeBindingResolvers;
 
 @{
     var typePart = Model.TypePartDefinition;
     var theme = await ThemeManager.GetThemeAsync();
     var shapeTable = ShapeTableManager.GetShapeTable(theme?.Id);
-    var displayShapes = shapeTable.Bindings.Keys.Where(x => x.StartsWith(Model.TypePartDefinition.PartDefinition.Name + "_DisplayOption__", StringComparison.OrdinalIgnoreCase) || x.Equals(Model.TypePartDefinition.PartDefinition.Name + "_DisplayOption", StringComparison.OrdinalIgnoreCase));
+
+    Func<string, bool> displayPredicate = (x) => (x.StartsWith(Model.TypePartDefinition.PartDefinition.Name + "_DisplayOption__", StringComparison.OrdinalIgnoreCase)
+                                  || x.Equals(Model.TypePartDefinition.PartDefinition.Name + "_DisplayOption", StringComparison.OrdinalIgnoreCase));
+
+    // Build query from Shape Table
+    var displayShapes = shapeTable.Bindings.Keys
+                            .Where(x => displayPredicate(x));
+
+    // Build query from Admin Templates
+    var resolverShapes = ShapeBindingResolvers
+                            .Select(async y => await y.GetShapeBindingNamesAsync(displayPredicate, true))
+                            .Where(c => c.Result != null)
+                            .SelectMany(t => t.Result);
```

Comment:
You can't use await like this in Linq. The `.Result` call here is blocking.

Better just to not use Linq in this scenario (at a pinch you can use System.Linq.Async, but not sure there is really any benefit to it here)

Reply:
re-written using foreach